contracts/src/GaspToken.sol

```diff
@@ -1,30 +1,168 @@
+    }
+
+    function whitelistSender(address sender) external override onlyOwner {
+        if (sender == address(0)) {
+            revert ZeroSender();
+        }
+
+        bool isAdded = _senderWhitelist.add(sender);
+        if (!isAdded) {
+            revert SenderAlreadyWhitelisted(sender);
+        }
+
+        emit SenderWhitelisted(sender);
+    }
+
+    function whitelistRecipient(address recipient) external override onlyOwner {
+        if (recipient == address(0)) {
+            revert ZeroRecipient();
+        }
+
+        bool isAdded = _recipientWhitelist.add(recipient);
+        if (!isAdded) {
+            revert RecipientAlreadyWhitelisted(recipient);
+        }
+
+        emit RecipientWhitelisted(recipient);
+    }
+
+    function dewhitelistSender(address sender) external override onlyOwner {
+        if (sender == address(0)) {
+            revert ZeroSender();
+        }
+
+        bool isRemoved = _senderWhitelist.remove(sender);
+        if (!isRemoved) {
+            revert SenderNotWhitelisted(sender);
+        }
 
-contract ArbToken is MintableERC20 {
-    constructor() MintableERC20("CrossChain Token", "CCXT") {}
-}
\ No newline at end of file
+        emit SenderDewhitelisted(sender);
+    }
+
+    function dewhitelistRecipient(address recipient) external override onlyOwner {
+        if (recipient == address(0)) {
+            revert ZeroRecipient();
+        }
+
+        bool isRemoved = _recipientWhitelist.remove(recipient);
+        if (!isRemoved) {
+            revert RecipientNotWhitelisted(recipient);
+        }
+
+        emit RecipientDewhitelisted(recipient);
+    }
+
+    function getSenderWhitelist() external view override returns (address[] memory) {
+        return _senderWhitelist.values();
+    }
+
+    function getRecipientWhitelist() external view override returns (address[] memory) {
+        return _recipientWhitelist.values();
+    }
+
+    function approve(address spender, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isApproveOperationForbidden(_msgSender(), spender)
+        returns (bool)
+    {
+        return super.approve(spender, amount);
+    }
+
+    function transfer(address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isTransferOperationForbidden(_msgSender(), recipient)
+        returns (bool)
+    {
+        return super.transfer(recipient, amount);
+    }
+
+    function transferFrom(address owner, address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isTransferFromOperationForbidden(owner, _msgSender(), recipient)
+        returns (bool)
+    {
+        return super.transferFrom(owner, recipient, amount);
+    }
+
+    function allowance(address owner, address spender) public view override(ERC20, IERC20) returns (uint256) {
+        return !allowTransfers && !_checkOwnerOrSpenderWhitelisted(owner, spender) ? 0 : super.allowance(owner, spender);
```

Comment:
The reason here is to turn off the allowance mechanism _completely_ for accounts which are not eligible to transfer during the initial phase.

E.g. consider a use theoretical case when some whitelisted account like the L1 council approves an account of user Alice to spend tokens. Due to the transfer restriction of the initial phase, Alice won't be able to transfer tokens anyway as she is not whitelisted. However, this method will return a non zero amount and it might confuse her.

So it's better to return zero here until the second phase is reached which enable transfers for anybody. And if an account is eligible to make transfers during the initial phase, it will see a non zero value if having been approved by a whiltelisted account.